File: vision/high_goal/high_goal.cpp
#include "high_goal.hpp"

#include <opencv2/core.hpp>
#include <opencv2/imgproc/imgproc.hpp>
#include <iostream>

// define to draw debugging information on images (bounding rectangles on
// targets)
#define DRAW_DEBUG

namespace frc::robot::vision {

// calculate angle from linear position on screen ( in radians)
double calcAngle(long value, long centerVal, double focalLen) {
  return atan((double) (value - centerVal) / focalLen);
}

const cv::Mat openKernel =
    getStructuringElement(cv::MORPH_RECT, cv::Size{3, 3}, cv::Point{-1, -1});

// number of contours to consider (by size)
const size_t numContours = 8;

// draw the outline of a rotated rectangle on image
void drawRotatedRect(cv::Mat &image, const cv::RotatedRect &rect,
                     const cv::Scalar &color) {
  cv::Point2f vertices2f[4];
  rect.points(vertices2f);

  cv::Point points[4];
  for (int i = 0; i < 4; ++i) {
    points[i] = vertices2f[i];
  }

  // draw lines between points
  for (int i = 0; i < 4; i++) {
    cv::line(image, points[i], points[(i + 1) % 4], color, 2);
  }
}

const double coverageExp = 1.5;
// score a contour based on what percentage of rectangle's area the given contour covers this score gives a measurement of how rectangular the contour is (since targets are very rectangular)
double areaCoverScore(const cv::RotatedRect &rect,
                      const std::vector<cv::Point> &contour) {
  auto coverage = cv::contourArea(contour) / abs(rect.size.area());
  return 1.0 - pow(1.0 - coverage, coverageExp);
}

// score a target on it's aspect ratio (should be close to 5"/2" = 2.5)
const double aspectExp = 0.8;
double aspectScore(const cv::RotatedRect &rect) {
  const double w = rect.size.width;
  const double h = rect.size.height;
  auto aspect = cv::max(w, h) / cv::min(w, h);

  return cv::max(1.0 - pow(abs(aspect - 2.6), aspectExp), 0.0);
}

double rotatedRectAngle(const cv::RotatedRect &rect) {
  return rect.size.width < rect.size.height ? rect.angle + 90.0 : rect.angle;
}

// score a target based on it's angle (should be near-ish horizontal)
double targetAngleScore(const cv::RotatedRect &rect) {
  double angle = rotatedRectAngle(rect);
  // expect angle to be in [0, 40] union [140, 180] (withing 40 degrees of horizontal)
  bool hit =
      (angle >= 0.0 && angle <= 40.0) || (angle >= 140.0 && angle <= 180.0);
  return hit ? 1.0 : 0.0;
}

// score a target based on it's relative size in the image
double targetSizeScore(const cv::RotatedRect &rect, double image_size, double relative_thresh) {
  double thresh = image_size * relative_thresh;
  return cv::max(rect.size.width, rect.size.height) >= thresh ? 1.0 : 0.0;
}

// check if a contour with the given bounding rectangle is a valid target
bool checkTarget(const cv::RotatedRect &rect,
                 const std::vector<cv::Point> &contour, double score_thresh, double image_size, double size_rel_thresh) {
  return (
            aspectScore(rect) +
            areaCoverScore(rect, contour) +
            targetAngleScore(rect) +
            2.0 * targetSizeScore(rect, image_size, size_rel_thresh)
          ) /
             5.0 >
         score_thresh;
}

std::optional<TargetAngle> find_target_angles(const VisionConfig& config, const CameraFOV &fov, cv::Mat &image) {
  // calculate camera information
  double aspectDiag = hypot(fov.aspect_h, fov.aspect_v);

  double fieldViewH = atan(tan(fov.diag_field_view / 2.0) * (fov.aspect_h / aspectDiag)) * 2.0;
  double fieldViewV = atan(tan(fov.diag_field_view / 2.0) * (fov.aspect_v / aspectDiag)) * 2.0;

  double hFocalLen = image.size().width / (2.0 * tan(fieldViewH / 2.0));
  double vFocalLen = image.size().height / (2.0 * tan(fieldViewV / 2.0));

  // convert to hsv
  cv::Mat mask;
  cv::cvtColor(image, mask, cv::COLOR_BGR2HSV);

  // generate hsv threshold mask
  cv::Scalar hsv_low(config.hsvLowH, config.hsvLowS, config.hsvLowV);
  cv::Scalar hsv_high(config.hsvHighH, config.hsvHighS, config.hsvHighV);

  cv::inRange(mask, hsv_low, hsv_high, mask);

  if(config.do_dilate) {
    cv::morphologyEx(mask, mask, cv::MORPH_DILATE, openKernel);
  }
  // close + open mask to patch up small holes
  cv::morphologyEx(mask, mask, cv::MORPH_CLOSE, openKernel, cv::Point(-1, -1), config.close_iters);
  cv::morphologyEx(mask, mask, cv::MORPH_OPEN, openKernel, cv::Point(-1, -1), config.open_iters);

  // find all contours in image
  std::vector<std::vector<cv::Point>> contours;
  std::vector<cv::Vec4i> hierarchy;
  cv::findContours(mask, contours, hierarchy, cv::RETR_EXTERNAL,
                   cv::CHAIN_APPROX_SIMPLE);

  // order contours by size
  std::sort(contours.begin(), contours.end(),
            [](std::vector<cv::Point> &c1, std::vector<cv::Point> &c2) -> bool {
              return abs(cv::contourArea(c1)) < abs(cv::contourArea(c2));
            });

  // filter and draw bounding rects
  std::vector<cv::RotatedRect> targets;
  for (size_t i = 0; i < cv::min(numContours, contours.size()); i++) {
    auto &contour = contours[i];
    auto rect = cv::minAreaRect(contour);
    auto hit = checkTarget(rect, contour, config.score_thresh, cv::max(image.size().width, image.size().height), config.size_rel_thresh);

    if (hit) {
#ifdef DRAW_DEBUG
      drawRotatedRect(image, rect, cv::Scalar(255.0, 0.0, 0.0));
#endif
      targets.push_back(rect);
    } else {
#ifdef DRAW_DEBUG
      drawRotatedRect(image, rect, cv::Scalar(0.0, 0.0, 255.0));
#endif
    }
  }

  // pick out highest point on contours
  int center_x = 0;
  int center_y = image.size().height;
  cv::RotatedRect *high_target = nullptr;
  for (auto &target : targets) {
    if (target.boundingRect().y < center_y) {
      // center x is taken to be the center of the rectangle + cos(angle) towards the higher side
      center_x = target.boundingRect().x + target.boundingRect().width * 0.5;
      auto angle = rotatedRectAngle(target);
      auto adjust = 0.5 * sin(angle * 3.1415926 / 180.0) * target.boundingRect().width;
      if(angle < 90.0) {
        center_x -= adjust;
      } else {
        center_x += adjust;
      }

      center_y = target.boundingRect().y;
      high_target = &target;
    }
  }

  if (high_target != nullptr) {
#ifdef DRAW_DEBUG
    // draw center on image
    cv::rectangle(image, cv::Point(image.size().width / 2, 0), cv::Point(image.size().width / 2, image.size().height), cv::Scalar(255.0, 255.0, 255.0));
    cv::rectangle(image, cv::Point(0, image.size().height / 2), cv::Point(image.size().width, image.size().height / 2), cv::Scalar(255.0, 255.0, 255.0));
    // draw target
    cv::circle(image, cv::Point(center_x, center_y), 5,
               cv::Scalar(0.0, 125.0, 255.0), 2);
#endif

    // calculate pitch and yaw angles
    double yaw = calcAngle(center_x, image.size().width / 2, hFocalLen);
    double pitch = -calcAngle(center_y, image.size().height / 2, vFocalLen);
    return TargetAngle { yaw, pitch };
  } else {
    return {};
  }
}

double get_distance_to_target(const TargetAngle& target, double target_height, double camera_angle) {
  return target_height / tan(target.pitch + camera_angle);
}

}

File: highgoal/high_goal.py
import cv2
import math
import numpy as np
from typing import List, Tuple
from numpy import ndarray
from .pipeline import get_filtered_contours
from .target_scoring import check_target
from .rectangle import Rect


def get_distance_to_target(target_pitch: float, target_height: float, camera_angle: float) -> float:
    """
    Get the forward distance from the camera to the target.

    :param target_pitch: the observed vertical angle between the camera and the target in radians
    :param target_height: the height of the target relative to the camera
    :param camera_angle: the vertical angle of the camera relative to the ground
    :return: the distance to the target in the same units as target_height
    """
    return target_height / math.tan(target_pitch + camera_angle)


def calc_angle(value: float, center_value: float, focal_length: float) -> float:
    """
    Calculate angle from linear position on screen (in radians)

    :param value: the value to calculate the angle to
    :param center_value: the value corresponding to the center of the screen
    :param focal_length: the focal length of the camera

    :return: the angle in radians to the value
    """
    return math.atan((value - center_value) / focal_length)


def find_target_angles(config: dict, image: np.ndarray) -> Tuple[float, float]:
    """
    Find the yaw and pitch of the best target within an image.

    :param config: the configuration dictionary
    :param image: the image to search for targets within
    :return: a tuple containing the yaw and pitch in radians
    """
    # unpack field of view information from config
    camera_fov = config['camera_fov']
    fov_h = camera_fov['aspect_h']
    fov_v = camera_fov['aspect_v']
    fov_diag = camera_fov['diag_field_view']

    # calculate camera information
    aspect_diagonal = math.hypot(fov_h, fov_v)
    field_view_h = math.atan(math.tan(fov_diag / 2.0) * (fov_h / aspect_diagonal)) * 2.0
    field_view_v = math.atan(math.tan(fov_diag / 2.0) * (fov_v / aspect_diagonal)) * 2.0

    h_focal_length = image.size().width / (2.0 * math.tan(field_view_h / 2.0))
    v_focal_length = image.size().height / (2.0 * math.tan(field_view_v / 2.0))

    # find targets
    score_thresh = config['vision_config']['score_thresh']
    size_relative_thresh = config['vision_config']['score_thresh']
    image_size = max(image.shape)
    image_center_x = image.shape[0] / 2.0
    image_center_y = image.shape[1] / 2.0

    targets = find_targets(config, image, score_thresh, image_size, size_relative_thresh)
    highest_target_rect = find_highest_target(image_center_y, targets)

    # calculate angles
    yaw = calc_angle(highest_target_rect.center_x, image_center_x, h_focal_length)
    pitch = -calc_angle(highest_target_rect.center_y, image_center_y, v_focal_length)

    return yaw, pitch


def find_targets(config: dict,
                 image: np.ndarray,
                 score_thresh: float,
                 image_size: float,
                 size_relative_thresh: float) -> List[Tuple[np.ndarray, Rect]]:
    """
    Return a list of targets given an image and some scoring parameters.

    :param config: the configuration parameters for the filtering and camera
    :param image: the image to locate targets within
    :param score_thresh: the minimum score a target must exceed to be returned
    :param image_size: the size of the image as its larger dimension
    :param size_relative_thresh: a proportion of image_size. See target_scoring.target_size_score

    :return: a list of targets as tuples of their contour and rectangle
    """
    contours = get_filtered_contours(config, image)
    targets = []
    for contour in contours:
        rect = Rect(cv2.boxPoints(cv2.minAreaRect(contour)))
        if check_target(rect, contour, score_thresh, image_size, size_relative_thresh):
            targets.append((contour, rect))
    return targets


def find_highest_target(center_y: float, targets: List[Tuple[np.ndarray, Rect]]) -> Rect:
    """
    Get the highest target from a list of targets.

    :param center_y: the y value of the center of the image
    :param targets: a list of targets, each represented with a tuple containing its contour and bounding rect

    :return: a rectangle around the target containing the highest point (or None if no targets are passed)
    """
    absolute_highest = None
    highest_target = None
    for target_contour, target_rect in targets:
        if target_rect.center_y > center_y:  # targets must be above middle of image or else they are false positives
            highest_in_contour = max(target_contour, key=lambda point: point[1])
            if absolute_highest is None or highest_in_contour[1] > absolute_highest[1]:
                absolute_highest = highest_in_contour
                highest_target = target_rect
    return highest_target
